### PR TITLE
Refactor to turn on `noUncheckedIndexedAccess` option for TypeScript

### PR DESCRIPTION
**Pull request by ybiquitous**
https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess

> Which issue, if any, is this issue related to?

Ref: https://github.com/stylelint/stylelint/pull/5982#issuecomment-1076966625

> Is there anything in the PR that needs further explanation?

The `noUncheckedIndexedAccess` option allows us to detect potential bugs like #5982, e.g.

```js
const firstNoe = node.nodes[0];
firstNode.value; // `firstNode` may be undefined, but TS compiler won't complain unless `noUncheckedIndexedAccess`.
```

Note: This pull request just helps us to see the errors if `noUncheckedIndexedAccess` is turned on. **We cannot merge it now.**

See https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess

---

EDIT: If you want to try it on a branch other than this one, you can use the CLI flag `--noUncheckedIndexedAccess`, for example:

```console
$ npm run lint:types -- --noUncheckedIndexedAccess
```

### CODE REVIEW

File: lib/rules/color-hex-length/index.js [hunk 1]
```diff
@@ -117,7 +117,7 @@ function longer(hex) {
 	let hexVariant = '#';
 
 	for (let i = 1; i < hex.length; i++) {
-		hexVariant += hex[i] + hex[i];
+		hexVariant += hex.charAt(i).repeat(2);
```
Comment (ybiquitous) [inline lib/rules/color-hex-length/index.js:120]: [note] [`.charAt()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/charAt) always returns a string (e.g. an empty string).

File: lib/rules/linebreaks/index.js [hunk 2]
```diff
@@ -60,9 +60,7 @@ const rule = (primary, _secondaryOptions, context) => {
 
 			const lines = root.source.input.css.split('\n');
 
-			for (let i = 0; i < lines.length; i++) {
-				let line = lines[i];
-
+			for (let [i, line] of lines.entries()) {
```
Comment (ybiquitous) [inline lib/rules/linebreaks/index.js:63]: [note] [`.entries()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/entries) makes a traditional `for` loop much readable.

File: lib/rules/selector-class-pattern/index.js [hunk 3]
```diff
@@ -110,16 +110,20 @@ const rule = (primary, secondaryOptions) => {
  * @returns {boolean}
  */
 function hasInterpolatingAmpersand(selector) {
-	for (let i = 0, l = selector.length; i < l; i++) {
-		if (selector[i] !== '&') {
+	for (const [i, char] of Array.from(selector).entries()) {
```
Comment (ybiquitous) [inline lib/rules/selector-class-pattern/index.js:113]: [note] [`Array.from()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from) can make a string to an array of characters.

**Comment by ybiquitous** [comment]
The CI is now all green. ✅

**Review by jeddy3 — APPROVED** [review]
This is fantastic work, thank you!

Having this option enabled will likely catch bugs that we've fallen foul to in the past.

**Comment by ybiquitous** [comment]
@jeddy3 Thank you for reviewing the big pull request! 😄

**Pull request merged**